exporter/uptraceexporter/exporter.go

```diff
@@ -0,0 +1,239 @@
+// limitations under the License.
+
+package uptraceexporter
+
+import (
+	"context"
+	"encoding/binary"
+	"time"
+
+	"github.com/uptrace/uptrace-go/spanexp"
+	"go.opentelemetry.io/collector/consumer/pdata"
+	"go.opentelemetry.io/otel"
+	"go.opentelemetry.io/otel/codes"
+	"go.opentelemetry.io/otel/label"
+	"go.opentelemetry.io/otel/metric"
+	"go.opentelemetry.io/otel/unit"
+	"go.uber.org/zap"
+)
+
+type traceExporter struct {
+	cfg    *Config
+	logger *zap.Logger
+	upexp  *spanexp.Exporter
+
+	backendLatency metric.Int64ValueRecorder
+	meterErr       error
+}
+
+func newTraceExporter(cfg *Config, logger *zap.Logger) (*traceExporter, error) {
+	upexp, err := spanexp.NewExporter(&spanexp.Config{
+		DSN: cfg.DSN,
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	meter := otel.Meter(
+		"github.com/open-telemetry/opentelemetry-collector-contrib/exporter/uptraceexporter")
+	backendLatency, err := meter.NewInt64ValueRecorder("uptrace.backend.duration",
+		metric.WithDescription("Uptrace backend latency"),
+		metric.WithUnit(unit.Milliseconds))
+	if err != nil {
+		otel.Handle(err)
+	}
+
+	exporter := &traceExporter{
+		cfg:    cfg,
+		logger: logger,
+		upexp:  upexp,
+
+		backendLatency: backendLatency,
+		meterErr:       err,
+	}
+
+	return exporter, nil
+}
+
+// pushTraceData is the method called when trace data is available.
+func (e *traceExporter) pushTraceData(ctx context.Context, traces pdata.Traces) (int, error) {
+	outSpans := make([]spanexp.Span, 0, traces.SpanCount())
+
+	rsSpans := traces.ResourceSpans()
+	for i := 0; i < rsSpans.Len(); i++ {
+		rsSpan := rsSpans.At(i)
+		resource := keyValueSlice(rsSpan.Resource().Attributes())
+
+		ils := rsSpan.InstrumentationLibrarySpans()
+		for j := 0; j < ils.Len(); j++ {
+			ilsSpan := ils.At(j)
+			lib := ilsSpan.InstrumentationLibrary()
+
+			spans := ilsSpan.Spans()
+			for k := 0; k < spans.Len(); k++ {
+				span := spans.At(k)
+
+				outSpans = append(outSpans, spanexp.Span{})
+				out := &outSpans[len(outSpans)-1]
+
+				out.ID = asUint64(span.SpanID().Bytes())
+				out.ParentID = asUint64(span.ParentSpanID().Bytes())
+				out.TraceID = span.TraceID().Bytes()
+
+				out.Name = span.Name()
+				out.Kind = spanKind(span.Kind())
+				out.StartTime = int64(span.StartTime())
+				out.EndTime = int64(span.EndTime())
+
+				out.Resource = resource
+				out.Attrs = keyValueSlice(span.Attributes())
+
+				out.StatusCode = statusCode(span.Status())
+				out.StatusMessage = span.Status().Message()
+
+				out.Events = uptraceEvents(span.Events())
+				out.Links = uptraceLinks(span.Links())
+
+				out.Tracer.Name = lib.Name()
+				out.Tracer.Version = lib.Version()
+			}
+		}
+	}
+
+	if len(outSpans) > 0 {
+		start := time.Now()
+		err := e.upexp.SendSpans(ctx, outSpans)
+		e.recordLatency(ctx, time.Since(start), err)
+		if err != nil {
+			return 0, err
+		}
+	}
+
+	return 0, nil
+}
+
+func (e *traceExporter) recordLatency(ctx context.Context, dur time.Duration, err error) {
+	if e.meterErr != nil {
```

Comment:
I am not too happy with this but I did not find a better way to check if `meter.NewInt64ValueRecorder` returned an error or not...